control-plane/csi-driver/node/src/node.rs

```diff
@@ -311,30 +316,61 @@ impl node_server::Node for Node {
         Ok(Response::new(NodeUnpublishVolumeResponse {}))
     }
 
-    /// Get volume stats method is currently not implemented,
-    /// although it's simple to do.
-    ///
-    /// TODO: Just read the data about capacity/used space
-    /// inodes/bytes from the system using the mountpoint.
+    /// Get volume stats method evaluates and returns capacity metrics
     async fn node_get_volume_stats(
         &self,
         request: Request<NodeGetVolumeStatsRequest>,
     ) -> Result<Response<NodeGetVolumeStatsResponse>, Status> {
         let msg = request.into_inner();
         trace!("node_get_volume_stats {:?}", msg);
-
-        /*
-        Ok(Response::new(NodeGetVolumeStatsResponse {
-            usage: vec![VolumeUsage {
-                total: 0 as i64,
-                unit: volume_usage::Unit::Bytes as i32,
-                available: 0,
-                used: 0,
-            }],
-        }))
-        */
-        error!("Unimplemented {:?}", msg);
-        Err(Status::new(Code::Unimplemented, "Method not implemented"))
+        if msg.volume_id.is_empty() {
+            return Err(failure!(
+                Code::InvalidArgument,
+                "Failed to stage volume: missing volume id"
+            ));
+        }
+        if msg.volume_path.is_empty() {
+            return Err(failure!(
+                Code::InvalidArgument,
+                "Failed to stage volume: missing volume path"
+            ));
+        }
+        debug!("Volume {} on {}", &msg.volume_id, &msg.volume_path);
+        // 1. Check if the path exists
+        // 2. Check if this is a blockDevice
+        // 3. if blockdevice, return a blank response
+        // 4. Get the volume NewMetricsStatFS
+        // 5. Get the metrics
+        let f = nix::sys::statfs::statfs(&*msg.volume_path);
+        match f {
+            Ok(info) => match info.filesystem_type() {
+                // TODO: use XFS_SUPER_MAGIC from libc when crate is upgraded
+                nix::sys::statfs::EXT4_SUPER_MAGIC | nix::sys::statfs::FsType(XFS_MAGIC) => {
```

Comment:
Why do we need this explicit check? This will make supporting other filesystem trickier. Can't we simply  return the usage below? This way when we support more filesystems this code won't need changing.

Reply:
I had the thought earlier too, I'll have to go back to my rough notes why I added it.